### PR TITLE
Rock Pi S review

### PR DESCRIPTION
**Pull request by frezik**

### CODE REVIEW

**Comment by dnmfarrell** [comment]
Thanks Timm, I will take a look!

File: content/article/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    "authors"     : ["timm-murray"],
+    "date"        : "2020-01-20T17:50:11",
+    "tags"        : [ "hardware", "linux", "gpio", "pi", "rockpi" ],
+    "draft"       : true,
+    "image"       : "/images/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library/rockpi.jpg",
+    "thumbnail"   : "",
+    "description" : "There are lots of single board computers out there, but not all of them have an simple library for getting input on a pin. Never fret, because most of them run Linux, and it's easy to access pins with just Linux.",
+    "categories"  : "hardware"
+  }
+
```
Comment (briandfoy) [inline content/article/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library.md]: This is looking really good. There are a few things that can get it better traction though.

I'd add a topic paragraph here. I'm not a Pi person and I don't know what GPIO is. A motivating case ("I was trying to do this, and need to do that") can draw people into the article. Imagine that the first two sentences are the only things people will see in previews on Twitter, and so on.

I see the technical details in this article, but what else are you doing? Making blinking lights, etc? A concrete, physical example would make this pop a bit more.

File: content/article/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library.md [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+Board Computers (SBCs), and one of them is the Rock Pi S, a new board from Seeed 
+Studio.
+
+In the interest of full disclosure, the Rock Pi S provided by Seeed 
+Studio for this review. Specifically, it's the model with 512MB of RAM and 
+4Gb of built-in flash.
+
+The Rock Pi S competes in the same segment as the Raspberry Pi Zero, particularly the Zero W 
+with built-in WiFi. Its form factor is different; where the Zero is 
+shaped like a stick of gum, the Rock Pi S is closer to a square. The Rock
```
Comment (briandfoy) [inline content/article/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library.md]: Can you take some pictures of your devices? I'd love to see those here.

File: content/article/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library.md [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+with built-in WiFi. Its form factor is different; where the Zero is 
+shaped like a stick of gum, the Rock Pi S is closer to a square. The Rock 
+also has USB-C for power, an ethernet jack, and a USB-A port. Depending on the 
+project, it can end up being cheaper than the Zero, since you don't have to 
+buy a micro-USB to USB-A adapter to hook up most other devices.
+
+You do still need a micro SD card. While there are versions of the Rock with 
+built-in flash, it's small and not meant for booting an OS. Note that the size of the built-in flash
+is listed in giga*bits*. The 4Gb version is actually 512 mega*bytes*. So get an 
+SD card, preferably one with an A1 rating, which specifies that it works well
```
Comment (briandfoy) [inline content/article/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library.md]: Ah, did not know this! Is there a resource you can link to for the right sort of card? Which particular one did you use?

File: content/article/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library.md [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+V12 board), so make sure you're looking at the right pinout for your board. 
+See the [GPIO header documentation](https://wiki.radxa.com/RockpiS/hardware/rockpiS#gpio).
+
+One thing to keep in mind is that the pins are specified as 3.3V input. That means you don't want to put 5V 
+into them. Feeding it 5V might cause glitches, or could break the board 
+completely.
+
+The sysfs interface from Linux is documented at:
+
+https://www.kernel.org/doc/html/latest/admin-guide/gpio/sysfs.html
```
Comment (briandfoy) [inline content/article/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library.md]: Make the a link, with a sentence like "Check out the [sysfs interface](...) ..."

File: content/article/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+processor, networking, a full sized USB jack, and running on Linux in a sub $20 
+price range. This might be a new favorite.
+
+```perl
+#!perl
+use v5.26;
+use warnings;
+use Time::HiRes 'sleep';
+
+use constant BASE_PATH => '/sys/class/gpio';
```
Comment (briandfoy) [inline content/article/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library.md]: Some File::Spec::Functions::cattle might be handy here, but not a big deal.

**Comment by frezik** [comment]
Addressed the comments above.  Fleshed out the sections on GPIO and SD card classes, added an image, and fixed up the sysfs link.

File: content/article/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library.md [hunk 6]
```diff
@@ -0,0 +1,212 @@
+
+  {
+    "title"       : "The Rock Pi S; or, How to Get GPIO Input Without a Library",
+    "authors"     : ["timm-murray"],
+    "date"        : "2020-01-20T17:50:11",
+    "tags"        : [ "hardware", "linux", "gpio", "pi", "rockpi" ],
+    "draft"       : true,
+    "image"       : "/images/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library/rockpi.jpg",
+    "thumbnail"   : "",
+    "description" : "There are lots of single board computers out there. GPIO (General Purpose Input Output) is what we usually use to see if something is on or off, but not all of these newer boards have a simple library for accessing them in Perl. Never fret, because most of them run Linux, and it's easy to access hardware pins with just Linux.",
```
Comment (dnmfarrell) [inline content/article/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library.md]: The site assumes descriptions are just a single sentence, apologies that wasn't obvious. I should add something about this to the contributor guide. Here's what happens with a long description:

image

How about breaking the title up into:

title: The Rock Pi S Review
description: How to get GPIO without a library
?

File: content/article/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library.md [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    "categories"  : "hardware"
+  }
+
+When writing articles like these, there is an inevitable comparison to the 
+Raspberry Pi series. There is no way to fight this, 
+and for good reason. The first Raspberry Pi ushered in a slew of Single 
+Board Computers (SBCs), and one of them is the Rock Pi S, a new board from Seeed 
+Studio.
+
+In the interest of full disclosure, the Rock Pi S provided by Seeed
```
Comment (dnmfarrell) [inline content/article/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library.md]: `s/S provided/S was provided/` I think?

File: content/article/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library.md (hunk 6 above)
Comment (frezik) [reply content/article/the-rock-pi-s-or-how-to-get-gpio-input-without-a-library.md]: Those changes look good, go ahead with them.

**Pull request merged**